Initial implementation margin-trim for block containers (without floats and margin collapsing).

sammygill opened the pull request:
#### 6bd29f7cff747f7bb1c9965e4b08e52099f204ee
<pre>
Initial implementation margin-trim for block containers (without floats and margin collapsing).
<a href="https://bugs.webkit.org/show_bug.cgi?id=249206">https://bugs.webkit.org/show_bug.cgi?id=249206</a>
rdar://103285722

Reviewed by Alan Baradlay.

Adds the  initial logic for margin-trim support in block containers when
there are no floats or collapsing margins. Support for both of these will
be done in their own patches.

A virtual method named shouldTrimChildMargin was added to help determine
if a box&apos;s block start or block end margin should be trimmed.
RenderBlockFlows defines its own version which checks a couple of things:
1.) That the box passed in is a block level box
2.) Whether the passed in MarginTrimType is set for the margin trim
property and if the child is either the first in flow child box (for
BlockStart) or the last in flow child box (for BlockEnd).

Spec reference: <a href="https://drafts.csswg.org/css-box-4/#margin-trim-block">https://drafts.csswg.org/css-box-4/#margin-trim-block</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-non-adjoining-item-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-non-adjoining-item.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-end-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-end.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-start-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block.html: Added.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::shouldTrimChildMargin const):
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeBlockDirectionMargins const):
(WebCore::RenderBox::constrainBlockMarginInAvailableSpaceOrTrim const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::shouldTrimChildMargin const):

Canonical link: <a href="https://commits.webkit.org/258457@main">https://commits.webkit.org/258457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66ce03053451b35e9d3ecb138dc25e0f92c525d6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11164 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111351 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106001 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2085 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109101 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107801 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92559 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24042 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4741 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25470 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4838 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10907 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44958 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5805 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6586 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3621,8 +3621,10 @@ void RenderBox::computeBlockDirectionMargins(const RenderBlock& containingBlock,
+    if (!containingBlock.shouldTrimChildMargin(MarginTrimType::BlockStart, *this))
+        marginBefore = minimumValueForLength(style().marginBeforeUsing(&containingBlockStyle), cw);
+    if (!containingBlock.shouldTrimChildMargin(MarginTrimType::BlockEnd, *this))
+        marginAfter = minimumValueForLength(style().marginAfterUsing(&containingBlockStyle), cw);
```
Should `marginBefore`/`marginAfter` be initialized to anything if we are doing a trim?

sammygill replied:
Yes actually we should make sure they are set to 0. Changed.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1356,6 +1356,16 @@ LayoutUnit RenderBlockFlow::collapseMarginsWithChildInfo(RenderBox* child, Rende
 
+bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const RenderBox& child) const
+{
+    auto blockChild = dynamicDowncast<RenderBlock>(child);
```
how does this work with block-level replaced boxes?

sammygill replied:
Ah I guess it would not work in that case. Perhaps instead I should try checking the display type of the child here

alanbaradlay replied:
The "Block" in RenderBlock is surely misleading and makes you think all block-level boxes end up being RenderBlocks (and that's one of the reasons why LFC ditched strong typing in layout tree). It's safer to consult the display type when  looking for blocks, inlines etc.

sammygill replied:
Changed this line to use `child.style().isDisplayBlockLevel() `

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -3647,8 +3647,14 @@ void RenderBox::computeBlockDirectionMargins(const RenderBlock& containingBlock,
+    if (containingBlock.shouldTrimChildMargin(MarginTrimType::BlockEnd, *this))
+        marginAfter = 0_lu;
+    else
+        marginAfter = minimumValueForLength(style().marginAfterUsing(&containingBlockStyle), cw);
```
I wonder if this pattern is common enough to warrant a helper function (especially with your follow up patches on grid/flexbox)

sammygill replied:
Yes I think so. I ended up needing something similar for the Flexbox implementation so I will add one

sammygill replied:
I added a new helper method, `constrainBlockMarginInAvailableSpaceOrTrim` let me know what you think. We may be able to move the computation of the available size to inside the method itself, but I took a quick look at my flexbox patch and it looks like there was a case where it was overridden to a different value.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1356,6 +1356,15 @@ LayoutUnit RenderBlockFlow::collapseMarginsWithChildInfo(RenderBox* child, Rende
+        return firstInFlowChildBox() == &child;
+    return lastInFlowChildBox() == &child;
+}
+
```
how does this work with non-adjoining margins? -we don't even end up here?

alanbaradlay replied:
also, I am wondering what happens to collapsed descendant margins  ("It also truncates any margins collapsed with such a margin.")

sammygill replied:
I don't have any test cases with non-adjoining margins (all of the container margins are 0). For the following example: 

```
<style>
.container {
    margin-block-start: 10px;
    padding-block-start: 10px;
    background-color: green;
    margin-trim: block-start;
}
.child {
    width: 100%;
    height: 100px;
    margin-block-start: 20px;
}
</style>

<container>
<child></child>
</container>
```

Do you think that the block-start margin of the child should not be trimmed since it is non-adjoining? I feel like it should be trimmed and set to 0px since the child's margin is still adjacent to the container's block-start edge (unless it means the margin edge).

I have another bug for collapsed margins so I was planning on dealing with them in another patch.

brentfulgham replied:
@sgill26 Can you adjust the title a bit to reflect that there are more steps to come related to collapsed margins, etc. Maybe call this "(Step 1)" or "(Initial Implementation)" or something to clarify to people looking at changelogs in the future that this isn't the entire set of changes needed.

sammygill replied:
Done

sammygill committed:
Initial implementation margin-trim for block containers (without floats and margin collapsing).
https://bugs.webkit.org/show_bug.cgi?id=249206
rdar://103285722

Reviewed by Alan Baradlay.

Adds the  initial logic for margin-trim support in block containers when
there are no floats or collapsing margins. Support for both of these will
be done in their own patches.

A virtual method named shouldTrimChildMargin was added to help determine
if a box's block start or block end margin should be trimmed.
RenderBlockFlows defines its own version which checks a couple of things:
1.) That the box passed in is a block level box
2.) Whether the passed in MarginTrimType is set for the margin trim
property and if the child is either the first in flow child box (for
BlockStart) or the last in flow child box (for BlockEnd).

Spec reference: https://drafts.csswg.org/css-box-4/#margin-trim-block

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-start-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-non-adjoining-item-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-non-adjoining-item.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-end-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-end.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-start-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-replaced-block.html: Added.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::shouldTrimChildMargin const):
* Source/WebCore/rendering/RenderBlockFlow.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::computeBlockDirectionMargins const):
(WebCore::RenderBox::constrainBlockMarginInAvailableSpaceOrTrim const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::shouldTrimChildMargin const):

Canonical link: https://commits.webkit.org/258457@main
